build-system/tasks/e2e/describes-e2e.js

```diff
@@ -204,23 +234,64 @@ class AmpPageFixture {
     capabilities.set('chromeOptions', chromeOptions);
 
     const builder = new Builder().withCapabilities(capabilities);
-    return builder.build().then(driver => {
-      const controller = new SeleniumWebDriverController(driver);
-      env.controller = controller;
-      env.ampDriver = new AmpDriver(controller);
-      this.driver_ = driver;
-    });
+    const driver = await builder.build();
+    const controller = new SeleniumWebDriverController(driver);
+    const ampDriver = new AmpDriver(controller);
+
+    env.controller = controller;
+    env.ampDriver = ampDriver;
+    this.driver_ = driver;
+
+    const {
+      testUrl,
+      experiments,
+      initialRect,
+    } = this.spec;
+    const {
+      environment,
+      // TODO(estherkim): browser
+    } = env;
+
+    if (Array.isArray(experiments)) {
+      await toggleExperiments(ampDriver, testUrl, experiments);
+    }
+
+    if (initialRect) {
+      const {width, height} = initialRect;
+      await controller.setWindowRect({width, height});
+    }
+
+    await ampDriver.navigateToEnvironment(environment, testUrl);
   }
 
   /** @override */
-  async teardown(unusedEnv) {
+  async teardown(env) {
+    const {controller} = env;
+    if (controller) {
+      await controller.switchToParent();
+    }
     if (this.driver_) {
       await this.driver_.quit();
     }
     this.driver_ = null;
   }
 }
 
+/**
+ * Toggle the given experiments for the given test URL domain.
+ * @param {!AmpDriver} ampDriver
+ * @param {string} testUrl
+ * @param {!Array<string>} experiments
+ * @return {!Promise}
+ */
+async function toggleExperiments(ampDriver, testUrl, experiments) {
+  await ampDriver.navigateToEnvironment(AmpdocEnvironment.SINGLE, testUrl);
+
+  for (const experiment of experiments) {
```

Comment:
I'll punt on this one, since at this point it strikes me as a premature optimization. We can address this later if it becomes a performance bottleneck over a large number of tests